src/sentry/lang/javascript/processor.py

```diff
@@ -281,67 +352,152 @@ def fetch_release_file(filename, release, dist=None):
+
+    If return value is not empty, the caller is responsible for closing the stream.
+    """
+    dist_name = dist and dist.name or None
+    releasefile_ident = ReleaseFile.get_ident(RELEASE_ARCHIVE_FILENAME, dist_name)
+    cache_key = get_release_file_cache_key(
+        release_id=release.id, releasefile_ident=releasefile_ident
+    )
+
+    result = cache.get(cache_key)
+
+    if result == -1:
+        return None
+    elif result:
+        return BytesIO(result)
+    else:
+        qs = ReleaseFile.objects.filter(
+            release=release, dist=dist, ident=releasefile_ident
+        ).select_related("file")
         try:
-            encoding = result[3]
+            releasefile = qs[0]
         except IndexError:
-            encoding = None
-        result = http.UrlResult(
-            filename, result[0], zlib.decompress(result[1]), result[2], encoding
-        )
+            # Cache as nonexistent:
+            cache.set(cache_key, -1, 60)
+            return None
+        else:
+            try:
+                file_ = fetch_retry_policy(lambda: ReleaseFile.cache.getfile(releasefile))
+            except Exception:
+                logger.error("sourcemaps.read_archive_failed", exc_info=sys.exc_info())
+
+                return None
+
+            # This will implicitly skip too large payloads.
+            cache.set(cache_key, file_.read(), 3600)
+            file_.seek(0)
+
+            return file_
+
+
+def compress(fp: IO) -> Tuple[bytes, bytes]:
+    """ Alternative for compress_file when fp does not support chunks """
+    content = fp.read()
+    return zlib.compress(content), content
+
+
+def fetch_release_artifact(url, release, dist):
+    """
+    Get a release artifact either by extracting it or fetching it directly.
+
+    If a release archive was saved, the individual file will be extracted
+    from the archive.
+
+    """
+    cache_key, cache_key_meta = get_cache_keys(url, release, dist)
+
+    result = cache.get(cache_key)
+
+    if result == -1:  # Cached as unavailable
+        return None
+
+    if result:
+        return result_from_cache(url, result)
+
+    start = time.monotonic()
+
+    release_file = fetch_release_archive(release, dist)
+    if release_file is not None:
+        try:
+            archive = ReleaseArchive(release_file)
+        except BaseException as exc:
+            logger.error("Failed to initialize archive for release %s", release.id, exc_info=exc)
+        else:
+            with archive:
+                try:
+                    fp, headers = get_from_archive(url, archive)
+                except KeyError:
+                    logger.debug(
+                        "Release artifact %r not found in archive (release_id=%s)", url, release.id
+                    )
+                    cache.set(cache_key, -1, 60)
+                except BaseException as exc:
+                    logger.error("Failed to read %s from release %s", url, release.id, exc_info=exc)
```

Comment:
If we hit this branch or the similar one 11 lines above, we effectively fall back to legacy fetching. It would be surprising if legacy files are available if there's also a release archive, even if it's broken. In the long run, we probably want to create an error cache entry just like we do in other error cases.

Reply:
👍  I'll add comments to remind us.